Yosemite/Yosemite/SnapshotsProvider/FetchResultSnapshotsProvider.swift

```diff
@@ -0,0 +1,156 @@
+    /// The conditions to use when fetching the results.
+    ///
+    /// In the future, we can allow this to be mutable if necessary.
+    private let query: Query
+
+    /// The publisher that emits snapshots when the fetch results arrive and when the data changes.
+    public var snapshot: AnyPublisher<FetchResultSnapshot, Never> {
+        snapshotSubject.eraseToAnyPublisher()
+    }
+    private let snapshotSubject = CurrentValueSubject<FetchResultSnapshot, Never>(FetchResultSnapshot())
+
+    private lazy var fetchedResultsController: NSFetchedResultsController<MutableType> = {
+        let fetchRequest = NSFetchRequest<MutableType>(entityName: MutableType.entityName)
+        fetchRequest.predicate = query.predicate
+        fetchRequest.sortDescriptors = [query.sortDescriptor]
+
+        let resultsController = storage.createFetchedResultsController(
+            fetchRequest: fetchRequest,
+            sectionNameKeyPath: query.sectionNameKeyPath,
+            cacheName: nil
+        )
+        resultsController.delegate = self
+        return resultsController
+    }()
+
+    public init(storageManager: StorageManagerType, query: Query) {
+        self.storage = storageManager.viewStorage
+        self.query = query
+    }
+
+    /// Start fetching and emitting snapshots.
+    public func start() throws {
+        try fetchedResultsController.performFetch()
+    }
+
+    /// Retrieve the immutable type pointed to by `objectID`.
+    ///
+    /// This is typically used in the `UITableViewDiffableDataSource`'s `CellProvider` in order to
+    /// display the row's information. Example:
+    ///
+    /// ```
+    /// private func makeCellProvider() -> UITableViewDiffableDataSource<String, FetchResultSnapshotObjectID>.CellProvider {
+    ///    return { [weak self] tableView, indexPath, objectID in
+    ///        guard let self = self, let cell = tableView.dequeueReusableCell(withIdentifier: ...) else {
+    ///            fatalError()
+    ///        }
+    ///
+    ///        let order = provider.object(withID: objectID)
+    ///        cell.configure(using: order)
+    ///        return cell
+    ///    }
+    ///}
+    /// ```
+    public func object(withID objectID: FetchResultSnapshotObjectID) -> MutableType.ReadOnlyType? {
+        assert(!objectID.isTemporaryID, "Expected objectID \(objectID) to be a permanent NSManagedObjectID.")
+
+        if let storageOrder = storage.loadObject(ofType: MutableType.self, with: objectID) {
+            return storageOrder.toReadOnly()
+        } else {
+            return nil
+        }
+    }
+
+    /// Converts the `NSFetchedResultsController` results to a snapshot that is then emitted
+    /// to observers.
+    ///
+    /// This will be hidden later. It should not be publicly accessible.
+    public func controller(_ controller: NSFetchedResultsController<NSFetchRequestResult>,
+                    didChangeContentWith snapshot: NSDiffableDataSourceSnapshotReference) {
```

Comment:
super cool that we don't need to implement those FRC delegate methods anymore! 🎉